src/main/java/jenkins/branch/NameMangler.java

```diff
@@ -0,0 +1,172 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.branch;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Locale;
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * Mangles names that are not nice so that they are safe to use on filesystem.
+ *
+ * @since 2.0.0
+ */
+public final class NameMangler {
+
+    private static final int MAX_SAFE_LENGTH = 32;
+    private static final int MIN_HASH_LENGTH = 6;
+    private static final int MAX_HASH_LENGTH = 12;
+
+    /**
+     * Utility class.
+     */
+    private NameMangler() {
+        throw new IllegalAccessError("Utility class");
+    }
+
+    public static String apply(String name) {
+        if (name.length() <= MAX_SAFE_LENGTH) {
+            boolean unsafe = false;
+            for (char c : name.toCharArray()) {
+                if (!isSafe(c)) {
+                    unsafe = true;
+                    break;
+                }
+            }
+            // See https://msdn.microsoft.com/en-us/library/aa365247 we need to consistently reserve names across all OS
+            if (!unsafe) {
+                // we know it is only US-ASCII if we got to here
+                switch (name.toLowerCase(Locale.ENGLISH)) {
+                    case ".":
+                    case "..":
+                    case "con":
+                    case "prn":
+                    case "aux":
+                    case "nul":
+                    case "com1":
+                    case "com2":
+                    case "com3":
+                    case "com4":
+                    case "com5":
+                    case "com6":
+                    case "com7":
+                    case "com8":
+                    case "com9":
+                    case "lpt1":
+                    case "lpt2":
+                    case "lpt3":
+                    case "lpt4":
+                    case "lpt5":
+                    case "lpt6":
+                    case "lpt7":
+                    case "lpt8":
+                    case "lpt9":
+                        unsafe = true;
+                        break;
+                    default:
+                        if (name.endsWith(".")) {
+                            unsafe = true;
+                        }
+                        break;
+                }
+            }
+            if (!unsafe) {
+                return name;
+            }
+        }
+        StringBuilder buf = new StringBuilder(name.length() + 16);
+        for (char c : name.toCharArray()) {
+            if (isSafe(c)) {
+                buf.append(c);
+            } else if (c == '/' || c == '\\' || c == ' ') {
+                buf.append('_');
+            } else if (c <= 0xff) {
+                buf.append('%');
+                buf.append(StringUtils.leftPad(Integer.toHexString(c & 0xff), 2, '0'));
```

Comment:
IIUC `%` itself is unsafe, so deliberately creating a branch name with a `%` in it will not cause problems—you will just wind up with `%25` in there. Right?